### DISCUSSION TITLE
Phase field modeling

### DISCUSSION THREAD

**Discussion by nishant1785** [idaholab/moose#22363] [Q&A General] [answered]
I am new to phase field modeling. I am running 3D version of this code & even after 1000 iterations it's not converging but in 2D it converges really quick. Parameters kappa and M were same for both cases (0.1 & 1). 

Thank you 

```
Test the non-split parsed function free enery Cahn-Hilliard Bulk kernel
# The free energy used here has the same functional form as the CHPoly kernel
# If everything works, the output of this test should replicate the output
# of marmot/tests/chpoly_test/CHPoly_test.i (exodiff match)
#

[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 16
  ny = 16
  xmax = 50
  ymax = 50
  elem_type = QUAD4
[]

[Variables]
  [./cv]
    order = THIRD
    family = HERMITE
  [../]
[]

[ICs]
  [./InitialCondition]
    type = CrossIC
    x1 = 5.0
    y1 = 5.0
    x2 = 45.0
    y2 = 45.0
    variable = cv
  [../]
[]

[Kernels]
  [./ie_c]
    type = TimeDerivative
    variable = cv
  [../]
  [./CHSolid]
    type = CahnHilliard
    variable = cv
    f_name = F
    mob_name = M
  [../]
  [./CHInterface]
    type = CHInterface
    variable = cv
    mob_name = M
    kappa_name = kappa_c
  [../]
[]

[Materials]
  [./consts]
    type = GenericConstantMaterial
    prop_names  = 'M kappa_c'
    prop_values = '1 0.1'
  [../]
  [./free_energy]
    type = DerivativeParsedMaterial
    f_name = F
    args = 'cv'
    function = '(1-cv)^2 * (1+cv)^2'
  [../]
[]

[Executioner]
  type = Transient
  scheme = 'bdf2'

  solve_type = 'NEWTON'
  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart'
  petsc_options_value = 'hypre boomeramg 31'

  l_max_its = 15
  l_tol = 1.0e-4
  nl_max_its = 10
  nl_rel_tol = 1.0e-11

  start_time = 0.0
  num_steps = 2
  dt = 0.7
[]

[Outputs]
  [./out]
    type = Exodus
    refinements = 1
  [../]
[]
```

**Comment by GiudGiud** [comment, marked as answer]
Hello

Can you try turning on automatic scaling on this?
Setting convergence parameters manually is not always a good idea, it can accept solution too early or never accept them if too tight.

Also could you please attach the console log?

Guillaume

**Reply by hugary1995** [reply]
MOOSE automatically changes the strong threshold for boomeramg depending on the topological dimension of your problem. IIRC the threshold is 0.4(or 0.25?) in 2D and 0.7 in 3D. As a result you may need to adjust some other parameters of boomeramg for it to work better. There is a nice documentation on boomeramg here: https://mooseframework.inl.gov/application_development/hypre.html

Another issue is that boomeramg doesn't work if you use the split formulation of cahn-hilliard. But it seems like you are solving the original fourth order pde, so that probably isn't culprit here.

**Reply by hugary1995** [reply]
You can also try a much stronger preconditioner, e.g. lu or ilu, just to see if it is a solver issue. If a direct solver works, then almost certainly you'll have to adjust the boomeramg parameters to make it stronger.

**Reply by GiudGiud** [reply]
Any update on this? @nishant1785

**Comment by friedmud** [comment]
An `nl_rel_tol` of `1-e11` is fairly strict!  I would loosen that up to `1e-8` or `1e-6`.  Can you post what your convergence history looks like in 2D and 3D?

**Comment by nishant1785** [comment]
Hello all, @friedmud and @GiudGiud i did decrease both tolerances you said and i also did automatic rescaling but still after 1000 iterations it did not converge, @hugary1995 for  lu or ilu preconditioner moose is saying no such object exists.
Here is input file and output file as well. Please take a look.
# #
[phase-field1.zip](https://github.com/idaholab/moose/files/9781780/phase-field1.zip)

**Reply by GiudGiud** [reply]
To set a lu precondition you need to pass:
```
petsc_options_iname = '-pc_type'
petsc_options_ivalue = 'lu'
```

**Reply by nishant1785** [reply]
thank you @GiudGiud  will do so and see.

**Reply by GiudGiud** [reply]
looks like it s the nonlinear iteration that does not converge well beyond 4e-4 ?

**Comment by nishant1785** [comment]
I forgot to mention it tok 2 days to reach 1000 iterations for 8 core processor PC. Does it take this long time ??

**Reply by friedmud** [reply]
No - it shouldn't take 1000 nonlinear iterations... ever.  If it's taking more than ~10 that's not great... more than ~50 means it's probably broken.

**Reply by friedmud** [reply]
Did you change your mesh between 2D and 3D?

**Comment by nishant1785** [comment]
Thanks for your response derek & Yeah i did change mesh

Nishant Gaikwad-7972836355
M. Tech student
Materials Department
DIAT, DRDO,Girinagar
411025-Pune

**Reply by friedmud** [reply]
Did you change the mesh density?  Or just the dimension?

**Reply by friedmud** [reply]
What I mean is - if it was 20x20 in 2D... did you make it 20x20x20 in 3D?

**Comment by nishant1785** [comment]
That's what I asked initially in this discussion that I took 2D program &
converted it to 3D but it's not converging at all. I have also attached
input file.

Nishant Gaikwad-7972836355
M. Tech student
Materials Department
DIAT, DRDO,Girinagar
411025-Pune

**Comment by nishant1785** [comment]
Sorry, Yeah yeah that's what i meant earlier I converted it to 3D that is
20*20*20

Nishant Gaikwad-7972836355
M. Tech student
Materials Department
DIAT, DRDO,Girinagar
411025-Pune

**Comment by nishant1785** [comment]
No update sir, it's still not converging.

Nishant Gaikwad-7972836355
M. Tech student
Materials Department
DIAT, DRDO,Girinagar
411025-Pune

**Reply by GiudGiud** [reply]
did you try what @hugary1995 suggested?

**Comment by nishant1785** [comment]
I did try stronger preconditioner but it didn't converge but I didn't try
boomeramg options. I will try that.

Nishant Gaikwad-7972836355
M. Tech student
Materials Department
DIAT, DRDO,Girinagar
411025-Pune

**Reply by GiudGiud** [reply]
Ok you should try svd preconditioning with the svd monitor. That will let us know if the problem is singular or not

This is very expensive so you ll need to recuse the number of dofs to a few thousands at most

**Comment by nishant1785** [comment]
But I am not getting this , openFoam software does this problem in 15
minutes but for MOOSE why is it so hard??

Nishant Gaikwad-7972836355
M. Tech student
Materials Department
DIAT, DRDO,Girinagar
411025-Pune

**Reply by GiudGiud** [reply]
The problem setup likely has issues. 
maybe @dschwen can comment on your inputs?
